### ISSUE TITLE
cloud_firestore 0.13.7 subtle crash when exit from the app and enter it again if save data is pending after exit.

### ISSUE THREAD

**Issue by asoftraiden** [FirebaseExtended/flutterfire#2912]
subtle crash when exit from the app and enter it again if save data is pending after exit.

code:

```
var forex = db.collection('forex').document('LPBPYm2NVCxcpb4YGpHo');
                                  var updateTime = FieldValue.serverTimestamp();
                                  await forex.updateData({
                                    'BuyUSD2IQD': double.parse(_buyTxt.text),
                                    'SellUSD2IQD': double.parse(_sellTxt.text),
                                    'maxUSD2IQD': double.parse(_maxTxt.text),
                                    'minUSD2IQD': double.parse(_minTxt.text),
                                    //'updateTime': Timestamp.fromDate(DateTime.now())
                                    'updateTime': updateTime
                                    // 'updateTime': Timestamp.now()
                                  })

Very Important note and fix:

I have a similar issue with firebase auth: [https://github.com/FirebaseExtended/flutterfire/issues/2886#issuecomment-654559796](url)

The fix was to downgrade to a pervious version mentioned in the link which is firebase_auth 0.15.5+3

The fix for this current issue is also to downgrade to the previous version of cloud_firestore version 0.13.4+2

I noted that whoever updated both is the same person(s) introduced this annoying subtle bug after April 2020








trace:

I/flutter ( 3512): document().snapshot().listen() callback in Constructor (document change detected)  ------------------------------------!
I/flutter ( 3512): setting state variables: ->>>>>>>>>>>>>>>>>>>>>>>>>>>>
I/flutter ( 3512): mounted = true >>>>> setting textEdit controllers
E/SpannableStringBuilder( 3512): SPAN_EXCLUSIVE_EXCLUSIVE spans cannot have a zero length
I/chatty  ( 3512): uid=10377(com.asoft.rayancontroller) identical 2 lines
E/SpannableStringBuilder( 3512): SPAN_EXCLUSIVE_EXCLUSIVE spans cannot have a zero length
W/ManagedChannelImpl( 3512): [{0}] Failed to resolve name. status={1}
I/flutter ( 3512): document().snapshot().listen() callback in Constructor (document change detected)  ------------------------------------!
I/flutter ( 3512): setting state variables: ->>>>>>>>>>>>>>>>>>>>>>>>>>>>
I/flutter ( 3512): mounted = true >>>>> setting textEdit controllers
W/rayancontrolle( 3512): Accessing hidden method Ldalvik/system/CloseGuard;->warnIfOpen()V (greylist,core-platform-api, linking, allowed)
I/flutter ( 3512): document().snapshot().listen() callback in Constructor (document change detected)  ------------------------------------!
I/flutter ( 3512): setting state variables: ->>>>>>>>>>>>>>>>>>>>>>>>>>>>
I/flutter ( 3512): mounted = true >>>>> setting textEdit controllers
W/ManagedChannelImpl( 3512): [{0}] Failed to resolve name. status={1}
I/flutter (  789): forex document is changed callback is called
I/flutter (  789): forex document is changed callback is called
I/flutter (  789): forex document is changed callback is called
I/flutter ( 3512): document().snapshot().listen() callback in Constructor (document change detected)  ------------------------------------!
I/flutter ( 3512): setting state variables: ->>>>>>>>>>>>>>>>>>>>>>>>>>>>
I/flutter ( 3512): mounted = true >>>>> setting textEdit controllers
E/SpannableStringBuilder( 3512): SPAN_EXCLUSIVE_EXCLUSIVE spans cannot have a zero length
E/SpannableStringBuilder( 3512): SPAN_EXCLUSIVE_EXCLUSIVE spans cannot have a zero length
E/SpannableStringBuilder( 3512): SPAN_EXCLUSIVE_EXCLUSIVE spans cannot have a zero length
E/SpannableStringBuilder( 3512): SPAN_EXCLUSIVE_EXCLUSIVE spans cannot have a zero length
E/SpannableStringBuilder( 3512): SPAN_EXCLUSIVE_EXCLUSIVE spans cannot have a zero length
E/SpannableStringBuilder( 3512): SPAN_EXCLUSIVE_EXCLUSIVE spans cannot have a zero length
E/SpannableStringBuilder( 3512): SPAN_EXCLUSIVE_EXCLUSIVE spans cannot have a zero length
E/SpannableStringBuilder( 3512): SPAN_EXCLUSIVE_EXCLUSIVE spans cannot have a zero length
E/SpannableStringBuilder( 3512): SPAN_EXCLUSIVE_EXCLUSIVE spans cannot have a zero length
I/chatty  ( 3512): uid=10377(com.asoft.rayancontroller) identical 2 lines
E/SpannableStringBuilder( 3512): SPAN_EXCLUSIVE_EXCLUSIVE spans cannot have a zero length
I/flutter ( 3512): document().snapshot().listen() callback in Constructor (document change detected)  ------------------------------------!
I/flutter ( 3512): setting state variables: ->>>>>>>>>>>>>>>>>>>>>>>>>>>>
I/flutter ( 3512): mounted = true >>>>> setting textEdit controllers
E/SpannableStringBuilder( 3512): SPAN_EXCLUSIVE_EXCLUSIVE spans cannot have a zero length
E/SpannableStringBuilder( 3512): SPAN_EXCLUSIVE_EXCLUSIVE spans cannot have a zero length
E/SpannableStringBuilder( 3512): SPAN_EXCLUSIVE_EXCLUSIVE spans cannot have a zero length
E/SpannableStringBuilder( 3512): SPAN_EXCLUSIVE_EXCLUSIVE spans cannot have a zero length
I/flutter (  789): forex document is changed callback is called
I/flutter ( 3512): document().snapshot().listen() callback in Constructor (document change detected)  ------------------------------------!
I/flutter ( 3512): setting state variables: ->>>>>>>>>>>>>>>>>>>>>>>>>>>>
I/flutter ( 3512): mounted = true >>>>> setting textEdit controllers
V/NativeCrypto( 3512): Read error: ssl=0x6eac8dc9c8: I/O error during system call, Software caused connection abort
V/NativeCrypto( 3512): Write error: ssl=0x6eac8dc9c8: I/O error during system call, Broken pipe
V/NativeCrypto( 3512): SSL shutdown failed: ssl=0x6eac8dc9c8: I/O error during system call, Success
W/ManagedChannelImpl( 3512): [{0}] Failed to resolve name. status={1}
W/ManagedChannelImpl( 3512): [{0}] Failed to resolve name. status={1}
W/ManagedChannelImpl( 3512): [{0}] Failed to resolve name. status={1}
W/ManagedChannelImpl( 3512): [{0}] Failed to resolve name. status={1}
W/ManagedChannelImpl( 3512): [{0}] Failed to resolve name. status={1}
W/ManagedChannelImpl( 3512): [{0}] Failed to resolve name. status={1}
I/flutter ( 3512): document().snapshot().listen() callback in Constructor (document change detected)  ------------------------------------!
I/flutter ( 3512): setting state variables: ->>>>>>>>>>>>>>>>>>>>>>>>>>>>
I/flutter ( 3512): mounted = true >>>>> setting textEdit controllers
E/SpannableStringBuilder( 3512): SPAN_EXCLUSIVE_EXCLUSIVE spans cannot have a zero length
I/chatty  ( 3512): uid=10377(com.asoft.rayancontroller) identical 2 lines
E/SpannableStringBuilder( 3512): SPAN_EXCLUSIVE_EXCLUSIVE spans cannot have a zero length
W/ManagedChannelImpl( 3512): [{0}] Failed to resolve name. status={1}
W/FlutterJNI( 3512): Tried to send a platform message to Flutter, but FlutterJNI was detached from native C++. Could not send. Channel: flutter/textinput. Response ID: 0
W/FlutterJNI( 3512): Tried to send a platform message to Flutter, but FlutterJNI was detached from native C++. Could not send. Channel: flutter/textinput. Response ID: 0
W/ManagedChannelImpl( 3512): [{0}] Failed to resolve name. status={1}
W/ActivityThread( 3512): handleWindowVisibility: no activity for token android.os.BinderProxy@f87fedc
E/SpannableStringBuilder( 3512): SPAN_EXCLUSIVE_EXCLUSIVE spans cannot have a zero length
E/SpannableStringBuilder( 3512): SPAN_EXCLUSIVE_EXCLUSIVE spans cannot have a zero length
I/flutter ( 3512): build() MyHomePageState: loggedIn=false => LoginScreen
I/flutter ( 3512): MyHomePageState onAuthStatusChanged with event = FirebaseUser(Instance of 'PlatformUser')
I/flutter ( 3512): finished registering firestore listen callback in the ForexEditorState constructor
I/flutter ( 3512): document().snapshot().listen() callback in Constructor (document change detected)  ------------------------------------!
I/flutter ( 3512): setting state variables: ->>>>>>>>>>>>>>>>>>>>>>>>>>>>
I/flutter ( 3512): mounted = true >>>>> setting textEdit controllers
I/flutter ( 3512): forex editor onAuthStatusChanged() with event =FirebaseUser(Instance of 'PlatformUser')
W/ManagedChannelImpl( 3512): [{0}] Failed to resolve name. status={1}
E/SpannableStringBuilder( 3512): SPAN_EXCLUSIVE_EXCLUSIVE spans cannot have a zero length
E/SpannableStringBuilder( 3512): SPAN_EXCLUSIVE_EXCLUSIVE spans cannot have a zero length
W/ActivityThread( 3512): handleWindowVisibility: no activity for token android.os.BinderProxy@da9f441
E/SpannableStringBuilder( 3512): SPAN_EXCLUSIVE_EXCLUSIVE spans cannot have a zero length
E/SpannableStringBuilder( 3512): SPAN_EXCLUSIVE_EXCLUSIVE spans cannot have a zero length
I/flutter ( 3512): build() MyHomePageState: loggedIn=false => LoginScreen
W/ManagedChannelImpl( 3512): [{0}] Failed to resolve name. status={1}
I/flutter ( 3512): MyHomePageState onAuthStatusChanged with event = FirebaseUser(Instance of 'PlatformUser')
I/flutter ( 3512): finished registering firestore listen callback in the ForexEditorState constructor
I/flutter ( 3512): document().snapshot().listen() callback in Constructor (document change detected)  ------------------------------------!
I/flutter ( 3512): setting state variables: ->>>>>>>>>>>>>>>>>>>>>>>>>>>>
I/flutter ( 3512): mounted = true >>>>> setting textEdit controllers
I/flutter ( 3512): forex editor onAuthStatusChanged() with event =FirebaseUser(Instance of 'PlatformUser')
W/IInputConnectionWrapper( 3512): getExtractedText on inactive InputConnection
W/IInputConnectionWrapper( 3512): getTextBeforeCursor on inactive InputConnection
W/IInputConnectionWrapper( 3512): getExtractedText on inactive InputConnection
W/IInputConnectionWrapper( 3512): getTextBeforeCursor on inactive InputConnection
W/IInputConnectionWrapper( 3512): getExtractedText on inactive InputConnection
W/IInputConnectionWrapper( 3512): getTextBeforeCursor on inactive InputConnection
D/AndroidRuntime( 3512): Shutting down VM
E/AndroidRuntime( 3512): FATAL EXCEPTION: main
E/AndroidRuntime( 3512): Process: com.asoft.rayancontroller, PID: 3512
E/AndroidRuntime( 3512): java.lang.NullPointerException: Attempt to invoke virtual method 'void io.flutter.plugin.common.MethodChannel.invokeMethod(java.lang.String, java.lang.Object)' on a null object reference
E/AndroidRuntime( 3512): 	at io.flutter.plugins.firebase.cloudfirestore.CloudFirestorePlugin$DocumentObserver.onEvent(CloudFirestorePlugin.java:429)
E/AndroidRuntime( 3512): 	at io.flutter.plugins.firebase.cloudfirestore.CloudFirestorePlugin$DocumentObserver.onEvent(CloudFirestorePlugin.java:402)
E/AndroidRuntime( 3512): 	at com.google.firebase.firestore.DocumentReference.lambda$addSnapshotListenerInternal$2(com.google.firebase:firebase-firestore@@21.3.0:504)
E/AndroidRuntime( 3512): 	at com.google.firebase.firestore.DocumentReference$$Lambda$3.onEvent(Unknown Source:6)
E/AndroidRuntime( 3512): 	at com.google.firebase.firestore.core.AsyncEventListener.lambda$onEvent$0(com.google.firebase:firebase-firestore@@21.3.0:42)
E/AndroidRuntime( 3512): 	at com.google.firebase.firestore.core.AsyncEventListener$$Lambda$1.run(Unknown Source:6)
E/AndroidRuntime( 3512): 	at android.os.Handler.handleCallback(Handler.java:883)
E/AndroidRuntime( 3512): 	at android.os.Handler.dispatchMessage(Handler.java:100)
E/AndroidRuntime( 3512): 	at android.os.Looper.loop(Looper.java:224)
E/AndroidRuntime( 3512): 	at android.app.ActivityThread.main(ActivityThread.java:7520)
E/AndroidRuntime( 3512): 	at java.lang.reflect.Method.invoke(Native Method)
E/AndroidRuntime( 3512): 	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:539)
E/AndroidRuntime( 3512): 	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:950)
E/MQSEventManagerDelegate( 3512): failed to get MQSService.
I/Process ( 3512): Sending signal. PID: 3512 SIG: 9
Lost connection to device.
```

**Comment by TahaTesser** [comment]
Hi @asoftraiden 
Can you please provide your `flutter doctor -v` your `flutter run --verbose` and a complete reproducible minimal code sample
Thank you

**Comment by Salakar** [comment]
Hey, thanks for the report. As part of our on-going work for #2582, this has been resolved in our Firebase Firestore rework (#2913) - which has now been merged into master. We'll look at publishing some prereleases in the next few days. Thank you
